src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.cs

```diff
@@ -0,0 +1,267 @@
+                var assignedNull = types.Remove(state.Void);
+
+                if (types.Count == 1)
+                {
+                    var toType = types.Single();
+                    if (assignedNull)
+                    {
+                        toType = toType.WithNullableAnnotation(Analyzer.Utilities.Lightup.NullableAnnotation.Annotated);
+                    }
+
+                    if (fromType.TypeKind == TypeKind.Interface)
+                    {
+                        if (!toType.AllInterfaces.Contains(fromType, SymbolEqualityComparer.Default))
+                        {
+                            // If the from type is an interface, the destination type must implement that interface
+                            // This deals with the case of going from T to Task<T> for example.
+                            return;
+                        }
+                    }
+                    else
+                    {
+                        if (!IsAncestorOf(fromType, toType))
+                        {
+                            // if the from type is a class, then the to type must be more derived than the
+                            // class
+                            return;
+                        }
+                    }
+
+                    if (toType.TypeKind == TypeKind.Class
+                        && !SymbolEqualityComparer.Default.Equals(fromType, toType)
+                        && toType.SpecialType != SpecialType.System_Object
+                        && toType.SpecialType != SpecialType.System_Delegate)
+                    {
+                        var fromTypeName = GetTypeName(fromType);
+                        var toTypeName = GetTypeName(toType);
+                        var diagnostic = Diagnostic.Create(desc, sym.Locations[0], sym.Name, fromTypeName, toTypeName);
+                        context.ReportDiagnostic(diagnostic);
+                    }
+                }
+            }
+
+            bool CanUpgrade(IMethodSymbol methodSym)
+            {
+                if (state.MethodsAssignedToDelegate.ContainsKey(methodSym))
+                {
+                    return false;
+                }
+
+                return true;
+            }
+
+            static string GetTypeName(ITypeSymbol type) => type.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat);
+        }
+
+        /// <summary>
+        /// Determines whether the current instance is an ancestor type of the parameter.
+        /// </summary>
+        /// <param name="potentialAncestor">The potential ancestor being inspected.</param>
+        /// <param name="potentialDescendant">The type to test.</param>
+        /// <returns><see langword="true"/> if <paramref name="potentialDescendant"/> derives directly or indirectly from <paramref name="potentialAncestor"/>.</returns>
+        private static bool IsAncestorOf(ITypeSymbol potentialAncestor, ITypeSymbol potentialDescendant)
```

Comment:
Can we use one of the below helper extension methods?

https://github.com/dotnet/roslyn-analyzers/blob/10632e6dfeb051173788f4e5cd616abeba6d9ff0/src/Utilities/Compiler/Extensions/ITypeSymbolExtensions.cs#L90-L135

https://github.com/dotnet/roslyn-analyzers/blob/10632e6dfeb051173788f4e5cd616abeba6d9ff0/src/Utilities/Compiler/Extensions/ITypeSymbolExtensions.cs#L44-L67

Reply:
Done.